### ISSUE TITLE
Mypy seems to associate the wrong stub file with the implementation file

### ISSUE THREAD

**Issue by nchammas** [python/mypy#11577]
**Bug Report**

I am seeing two issues that are potentially related.

1. Mypy does not flag the same errors when pointed at a directory vs. when pointed at the individual files in the directory.
2. Mypy seems to associate the incorrect stub file with a given implementation file.

**To Reproduce**

I hit these issues while contributing to the Apache Spark project. To reproduce them, first clone the project and install their requirements:

```sh
# I was working at commit 3d0663c45f900332ff1ea826f45cbf2c79eba3cd.
git clone --depth=1 git@github.com:apache/spark.git
cd spark/python
python -m venv venv
source venv/bin/activate
pip install -r ../dev/requirements.txt
```

Now run mypy against the `pyspark/ml/` directory, and against an individual file in that same directory:

```sh
mypy --version  # mypy 0.910
mypy --config mypy.ini pyspark/ml/  # Success: no issues found in 38 source files
mypy --config mypy.ini pyspark/ml/common.py  # Found 6 errors in 1 file (checked 1 source file)
```

For some reason, `common.py` has several problems that are not identified when running mypy against the whole `ml/` directory. That is the first issue.

The second issue is even more confusing. In trying to understand what was going on, I introduced a syntax error into the stub file of an identically named implementation but in a different directory:

```sh
echo "\n234syntaxERROR\n" >> pyspark/mllib/common.pyi
mypy --config mypy.ini pyspark/ml/common.py  # pyspark/mllib/common.pyi:37: error: unexpected character after line continuation character  [syntax]
```

So for some reason mypy is matching up `pyspark/ml/common.py` with `pyspark/mllib/common.pyi`. Note the difference between `ml/` and `mllib/`. And you can see the same thing in reverse, too: Mypy matches up `pyspark/mllib/common.py` with `pyspark/ml/common.pyi`.

**Expected Behavior**

1. If `mypy dir/` returns no errors, I expect `mypy dir/somefile.py` to also return no errors. If `mypy dir/somefile.py` returns errors, I expect `mypy dir/` to include those same errors.
2. I expect Mypy to match up `ml/common.py` with `ml/common.pyi`, and `mllib/common.py` with `mllib/common.pyi`.

**Actual Behavior**

See the reproduction above.

**Your Environment**

- Mypy version used: mypy 0.910
- Mypy command-line flags: `--config mypy.ini`
- Mypy configuration options from `mypy.ini` (and other config files): https://github.com/apache/spark/blob/3d0663c45f900332ff1ea826f45cbf2c79eba3cd/python/mypy.ini
- Python version used: Python 3.9.8
- Operating system and version: macOS 11.6.1

#9954 is potentially related, but the Apache Spark project does not set `follow_imports` to a non-default value.

**Comment by nchammas** [comment]
cc @zero323

**Comment by erictraut** [comment]
Why are you creating stub files by the same name as implementation files?

A stub file is not "merged" with the implementation. It is meant to be a stand-in for the implementation, a way to document the interface of the implementation for _external_ consumers of a module.

If your intent is to type check the source code, type annotations should be added inline within the source. Creating type stubs with the same names as the source files will result in odd behaviors like those you're describing if you run a type checker on the code.

For example, when you ask mypy to type check `common.py`, it will open that source file and analyze the code within. When it encounters an import statement, it will resolve it using a type stub if a stub is present. If `common.py` imports from another source file, which in turn imports from `common`, the second import will resolve to `common.pyi`. If `common.py` and `common.pyi` both declare a class with the same name, these classes are now distinct from the perspective of the type checker. They aren't somehow "merged" into a single class definition.

In summary, if you are creating a type stub file next to a source file of the same name, you're probably doing something wrong, and you're going to experience a lot of problems if you then try to run a type checker on the code.

**Comment by nchammas** [comment]
I didn't create these files. They are part of the existing Apache Spark project and were contributed by others to that project.

So for problem 2, are you saying the "mismatched" stub files are somehow caused by an import inside `ml/common.py` leading to `mllib/common.pyi` (and vice-versa, an import inside `mllib/common.py` leading to `ml/common.pyi`)? I guess it's theoretically possible, though I'm not seeing an import chain that leads from one to the other.

Also, I'm not seeing how this explains problem 1, where pointing mypy at the `ml/` directory containing `common.py` does not identify any errors, but pointing it directly at the file does.

Anyway, I will eliminate both `ml/common.pyi` and `mllib/common.pyi` (by merging them into the files they are intended to describe) and see if that resolves these problems, and then report back.

**Comment by zero323** [comment]
@erictraut

> In summary, if you are creating a type stub file next to a source file of the same name, you're probably doing something wrong, and you're going to experience a lot of problems if you then try to run a type checker on the code.

I wonder if you could provide some reference for this statement. 

Both [PEP 484](https://www.python.org/dev/peps/pep-0484/#toc-entry-37)


> The easiest form of stub file storage and distribution is to put them alongside Python modules in the same directory. This makes them easy to find by both programmers and the tools


and [mypy docs](https://mypy.readthedocs.io/en/stable/stubs.html?highlight=stub%20files#creating-a-stub)

> Write a stub file for the library (or an arbitrary module) and store it as a .pyi file in the same directory as the library module.

clearly describe storing stub files alongside annotated module as both valid and preferred approach. 

> For example, when you ask mypy to type check common.py, it will open that source file and analyze the code within. When it encounters an import statement, it will resolve it using a type stub if a stub is present. 

Could you elaborate on this, please? Maybe I misunderstood things, but this seems directly contradict [PEP 561  Type Checker Module Resolution Order](https://www.python.org/dev/peps/pep-0561/#partial-stub-packages)

> If your intent is to type check the source code,

To clarify ‒ that was not the intention. Stub files, in the current form are intended for end users, not to type check the code.

**Comment by nchammas** [comment]
OK, eliminating `ml/common.pyi` and `mllib/common.pyi` resolves both of my issues. Thank you for the clear explanation and correct diagnosis, @erictraut.

I will close this issue as "user error", though I wish mypy would be clearer about its behavior in situations like this. The verbose output was very large and didn't show me how, for example, `ml/common.py` led to `mllib/common.pyi`, or why `mypy ml/common.py` flagged errors but `mypy ml/` didn't. I don't know enough about mypy to offer a specific debugging or observability suggestion.

@zero323, you probably already knew this, but I guess my mental model of mypy needed correction. I assumed all the `.pyi` files were contributing to our Python lint checks, but it seems we need to merge those files into the implementations they describe if we want the lint checks to use them. We can discuss this over on the dev list or something.

**Comment by erictraut** [comment]
@zero323, apologies if my previous answer appeared to contradict other documentation. I should have been more precise. PEP 484 and the mypy docs you cited are referring to a different situation — one where the maintainer of a library has no desire to use a static type checker on the library's implementation but still wants to provide type information for consumers of that library. From what you said, it sounds like this is the intent here. In that case, supplying "sidecar" type stubs alongside the source files is a supported solution.

If the intent is to run a static type checker on the library's source files and take advantage of static type checking to catch bugs (e.g. as part of a CI script), then "sidecar" type stub files are not appropriate and will cause the sort of problems that @nchammas encountered.

If you're a maintainer of a published Python library, you might want to consider moving away from "sidecar" type stubs and switch to inlined type annotations. The latter provide three major benefits:

First, there's overhead in keeping the type stubs and the implementation in sync. It's easy for the contents of these files to diverge, and consumers of the library suffer when the type information is inaccurate. Inlined type annotations are typically easier to maintain.

Second, "sidecar" stubs tends to degrade the experience for consumers of the library because the type information is separated from other useful information like doc strings and default argument values. That means that language servers and smart editors need to do their best to associate the type information (which comes from the type stubs) with the other details (which come from the implementation). While this is possible much of the time, it involves a bunch of complex heuristics that sometimes fail. (Full disclosure: I'm a contributor to Microsoft's Pyright type checker and Pylance language server, and I spend a lot of time thinking about how to improve the development experience for Python developers.)

Third, with inlined type annotations, you can leverage static type checking to catch bugs in the library code (e.g. as part of CI scripts).

Static typing for Python has come a long way since the early days of PEP 484 and mypy, and some of the older documentation doesn't reflect the latest guidance and recommendations. There has been a push recently to consolidate the latest Python typing documentation in the [python/typing](https://github.com/python/typing) github site. In particular, library maintainers might find [this document](https://github.com/python/typing/blob/master/docs/source/libraries.rst) of interest.

**Comment by zero323** [comment]
Thanks for clarification and all the suggestions  @erictraut.

Long story short ‒ the current state of things is a result of quite a few years of independent development and step-wise migration on large codebase with complex (JVM libraries for starters) dependencies and  conservative policies. So while in general I agree with some of your points, I am pretty sure that at least some parts of our annotations won't be inlined in forseable future.

**Comment by zero323** [comment]
Thanks for clarification and all the suggestions  @erictraut.

Long story short ‒ the current state of things is a result of quite a few years of independent development and step-wise migration on large codebase with complex (JVM libraries for starters) dependencies and  conservative policies. So while in general I agree with some of your points, I am pretty sure that at least some parts of our annotations won't be inlined in foreseeable future.
